### PR TITLE
feat: Retrieve finalized blocks from Ethereum

### PR DESCRIPTION
**Pull request by bvrooman**
Related issues:
- Closes https://github.com/FuelLabs/fuel-core/issues/1336

This PR changes the approach a Fuel node uses to determine the finalized blocks on the Ethereum blockchain.

**The existing approach:**
- Assume a "finalization period" - a period of time after which we assume a block is finalized. Generally, we assume a block is finalized after 100 new blocks have been produced afterwards
- Observe what the current block height is as reported by the Ethereum client
- Subtract the finalization period from the current height to get an assumed finalized block, i.e. current block height - 100

The problem with this approach, as outlined by Trail of Bits, is that finality cannot be guaranteed simply by the amount of time passed alone. 

**This PR proposes a more robust approach:**
- Use the Ethereum client to query for the most recently finalized block directly. This is done using the `Finalized` tag in the query. See https://github.com/gakonst/ethers-rs/pull/1792

This approach removes any assumption on our side as to what the finalized block is, and gets the canonically accepted finalized block from Ethereum.

### CODE REVIEW

File: crates/services/relayer/src/service/state.rs [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 use core::ops::RangeInclusive;
 pub use state_builder::*;
-use std::ops::Deref;
+// use std::ops::Deref;
```
Comment (Voxelot) [inline crates/services/relayer/src/service/state.rs:6]:
```suggestion
```

File: crates/services/relayer/src/service/state.rs [hunk 2]
```diff
@@ -53,50 +49,34 @@ impl EthState {
...
 
     /// Get the gap between the relayer and the Ethereum node if
     /// a sync is required.
     pub fn needs_to_sync_eth(&self) -> Option<EthSyncGap> {
         (!self.is_synced()).then(|| {
-            EthSyncGap::new(
-                self.local.map(|l| l.saturating_add(1)).unwrap_or(0),
-                self.remote.finalized(),
-            )
+            let local = self.local.map(|l| l.saturating_add(1)).unwrap_or(0);
```
Comment (Voxelot) [inline crates/services/relayer/src/service/state.rs:59]: Does removing the use of an inclusive range affect the correctness of doing a saturating_add here? It seems like there's a potential that local could be higher than remote in this case.

Comment (bvrooman) [reply crates/services/relayer/src/service/state.rs:58]: I do not believe that this changes the correctness.  Prior to these changes, `EthSyncGap` used a single inclusive range to store its start and end values (`local` and `remote` respectively); with these changes, `EthSyncGap` stores its start and end values separately, and implements a conversion into an inclusive range where any range behaviour needs to be maintained. In effect, the correctness will be maintained.

At the point of instantiation, where we calculate the inputs to the `EthSyncGap` constructor, `local` remains unchanged, and `remote` is now the value returned by the Ethereum client. Before and after these changes, there is no guarantee that `local` was less than or equal to `remote`. This means that the `EthSyncGap` can be "empty", i.e., `local` is greater than or equal to `remote`. In fact, some tests check the behaviour when `local` is greater than `remote` to ensure that no sync is necessary: https://github.com/FuelLabs/fuel-core/blob/ec6a61277118d2b9e3deb9e29ae01f8129242ea8/crates/services/relayer/src/service/state/test.rs#L30

File: tests/tests/relayer.rs [hunk 3]
```diff
@@ -309,6 +309,10 @@ async fn handle(
             let r = mock.get_block_number().await.unwrap();
             json!({ "id": id, "jsonrpc": "2.0", "result": r })
         }
+        "eth_getBlockByNumber" => {
```
Comment (xgreenx) [inline tests/tests/relayer.rs:312]: I guess we can remove `eth_blockNumber` now

**Review by xgreenx — APPROVED** [review]

File: tests/tests/relayer.rs (hunk 3 above)
Comment (bvrooman) [reply tests/tests/relayer.rs:308]: Yep, that's true. If we need it again in the future for whatever reason, we can add it back.

**Review by xgreenx — APPROVED** [review]

**Pull request merged**